data-reconciliation-app/docs/adr/05-Reporting.md

```diff
@@ -0,0 +1,115 @@
+# Data Reporting API
+
+## Status
+
+Proposed
+
+## Context
+
+We need to build an API that reports the members' reconciled dataset back to them.
+
+## Requirements:
+- A report can be requested for a single or for all records.
+- Members will receive a report only for data they have ingested.
+- A reconciliation report can be issued at anytime, by any member, and the report result will be presented based on the most recent data present in the KV-store for the entire network.
+- Rules for reconciliation are described in [Reconciliation Logic ADR](https://github.com/microsoft/ccf-app-samples/blob/main/data-reconciliation-app/docs/adr/03-reconciliation-logic.md).
+- If a report turns out to be empty, an Error message shall be returned saying that user has no data to be reconciled 
+
+## API Endpoints
+
+### Report for a single record
+
+- Description
+  - Report a single record by getting the key from the KV-store and run reconciliation logic
+- Path
+  - /report/{key}
+- HTTP Method
+  - GET
+- URL Params
+  - key: the record key to be found in the KV-store
+- Headers
+  - N/A
+- Request will be a `ccfapp.Request Object`. We will interrogate the `ccfapp.Request Object` to extract the user or member and [authenticate them via certficates](#security). We will read the request body as a JSON.
+- API Status Codes
+  - OK
+    - Status: 200
+    - Response: SummaryResult object in JSON format
+  - UNAUTHORIZED
+    - Status: 401
+    - Response: "Unauthorized"
+  - NO_CONTENT
+    - Status: 400
+    - Response: "Key does not exist."
+
+
+### Report for all data
+
+- Description
+  - Report all data scanning the KV-store for keys a user has submited data.
+- Path
+  - /report
+- HTTP Method
+  - GET
+- URL Params
+  - N/A
+- Headers
+  - N/A
+- Request will be a `ccfapp.Request Object`. We will interrogate the `ccfapp.Request Object` to extract the user or member and [authenticate them via certficates](#security). We will read the request body as a JSON.
+- API Status Codes
+  - OK
+    - Status: 200
+    - Response: Array of SummaryResult objects in JSON format
+  - UNAUTHORIZED
+    - Status: 401
+    - Response: "Unauthorized"
+  - NO_CONTENT
+    - Status: 204
+    - Response: "No data to report."
+
+
+### Security
+
+Users will be authenticated via certificates, which is natively supported by the CCF framework.
+
+//TODO Add reference to this future ADR: https://github.com/microsoft/ccf-app-samples/issues/102
+
+
+## Service
+
+### Reporting Service
+
+The reporting service will take the record(s) from KV-store and create SummaryResult object(s).
+
+- **For all data:** This service will be responsible to iterate the KV-store checking whether the member has submitted data against the key or not.
```

Comment:
This is going to be an expensive operation. Is this only for the demo purpose?
Apologies, if it has already been discussed earlier

Reply:
Definitely going to be expensive...we have discussed this and noted it as a conseuqence to our current repository design: https://github.com/microsoft/ccf-app-samples/blob/main/data-reconciliation-app/docs/adr/01-data-ingest.md#consequences

I added a story to the backlog to investigate our repository model and see if there is a better design to improve performance: https://github.com/orgs/microsoft/projects/542/views/1

Good observation! @kusweta

Reply:
Yep. I'm not happy with that either but that's how it will be for now :/